### PR TITLE
[BALDE-74] Fix input system

### PR DESCRIPTION
**Pull request by Blinkuu**

### CODE REVIEW

File: engine/src/core/events/event_manager.h [hunk 1]
```diff
@@ -45,45 +45,45 @@ namespace Bald {
...
         template<class T>
         void Unsubscibe(unsigned id) noexcept;
 
         /**
-        * @fn                    Emit
-        * @brief                 Templated function which is used to emit events of any kind
-        * @param [Args&& ...]    args -> Arguments for event's constructor (can be left blank)
+        * @fn Emit
+        * @brief Templated function which is used to emit events of any kind
+        * @param [Args& ...]    args -> Arguments for event's constructor (can be left blank)
```
Comment (BIGbadEL) [inline engine/src/core/events/event_manager.h]: why you deleted & from type?

File: engine/src/core/input/input_manager.cpp [hunk 2]
```diff
@@ -41,9 +41,9 @@ namespace Bald::Input {
         std::memcpy(m_MouseButtonsState.begin(), m_MouseButtons.begin(), MAX_MOUSE_BUTTONS * sizeof(m_MouseButtons[0]));
     }
 
-    void InputManager::EmitKeyPressedEvent(unsigned keycode) noexcept {
+    void InputManager::EmitKeyPressedEvent(unsigned keycode, bool isRepeated) noexcept {
         if(keycode >= MAX_KEYS) CORE_LOG_WARN("[InputManager] Wrong key id");
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.cpp]: I think you could change it to assert

File: engine/src/core/input/input_manager.cpp [hunk 3]
```diff
@@ -56,14 +56,9 @@ namespace Bald::Input {
         else Bald::EventManager::Emit<KeyReleasedEvent>(keycode);
     }
 
-    void InputManager::EmitMouseButtonPressedEvent(unsigned buttoncode) noexcept {
+    void InputManager::EmitMouseButtonPressedEvent(unsigned buttoncode, bool isRepeated) noexcept {
         if(buttoncode >= MAX_MOUSE_BUTTONS) CORE_LOG_WARN("[InputManager] Wrong mouse button id");
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.cpp]: Same as above

File: engine/src/core/input/input_manager.h [hunk 4]
```diff
@@ -272,21 +256,25 @@ namespace Bald::Input {
         }
     }
 
+    inline void char_callback([[maybe_unused]] GLFWwindow* window, unsigned character) {
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h:180]: where this function is used?

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: engine/src/core/events/event_manager.h (hunk 1 above)
Comment (Blinkuu) [reply engine/src/core/events/event_manager.h]: missclick

File: engine/src/core/input/input_manager.cpp (hunk 2 above)
Comment (Blinkuu) [reply engine/src/core/input/input_manager.cpp]: I don't think that's necessary. If somebody uses some weird type input device, we don't want to halt the engine.

File: engine/src/core/input/input_manager.cpp (hunk 3 above)
Comment (Blinkuu) [reply engine/src/core/input/input_manager.cpp]: Answered above.

File: engine/src/core/input/input_manager.h (hunk 4 above)
Comment (Blinkuu) [reply engine/src/core/input/input_manager.h:180]: This will be used in Window class.

File: engine/src/core/input/input_manager.h [hunk 5]
```diff
@@ -272,21 +256,25 @@ namespace Bald::Input {
...
+        InputManager::EmitKeyTypedEvent(character);
+    }
+
     inline void
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: It looks strange in two lines.

File: engine/src/core/input/input_manager.h [hunk 6]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
-
-        static void EmitMouseButtonTypedEvent(unsigned buttoncode) noexcept;
-
-        /**
-         * @fn                  EmitMouseButtonPressedEvent
-         * @brief               Emits MouseButtonPressedEvent
-         * @param keycode [int] GLFW key id
+         * @fn EmitMouseButtonPressedEvent
+         * @brief Emits MouseButtonPressedEvent
+         * @param [int] keycode -> GLFW key id
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: There is no documentation about the second parameter.

File: engine/src/core/input/input_manager.h [hunk 7]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
-
-        /**
-         * @fn                     SetMouseButtonReleasedCallback
-         * @brief                  sets function which will be called on mouse button released event
-         * @param fun              function pointer
-         * @param args             arguments for function
+         * @fn SetMouseButtonReleasedCallback
+         * @brief sets function which will be called on mouse button released event
+         * @param buttoncode [int] GLFW mouse button id
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: probably should mansion that function has to take Evnet& as the first parameter.

File: engine/src/core/input/input_manager.h [hunk 8]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
 
         /**
-         * @fn                     SetMouseButtonPressedCallback
-         * @brief                  sets function which will be called on mouse button pressed event
+         * @fn SetMouseButtonPressedCallback
+         * @brief sets function which will be called on mouse button pressed event
          * @param buttoncode [int] GLFW mouse button id
-         * @param fun              function pointer
-         * @param args             arguments for function
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: probably should mansion that function has to take Evnet& as the first parameter.

File: engine/src/core/input/input_manager.h [hunk 9]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
         static void SetMouseMovedCallback(F&& fun, Args&& ... args) noexcept;
 
         /**
-         * @fn                     SetMouseScrolledCallback
-         * @brief                  sets function which will be called on mouse scrolled event
-         * @param fun              function pointer
-         * @param args             arguments for function
+         * @fn SetMouseScrolledCallback
+         * @brief sets function which will be called on mouse scrolled event
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: probably should mansion that function has to take Evnet& as the first parameter.

File: engine/src/core/input/input_manager.h [hunk 10]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
         inline static void GetMousePos(double& xpos, double& ypos) noexcept;
 
         /**
-         * @fn                  SetKeyPressedCallback
-         * @brief               sets function which will be called on key pressed event
-         * @param fun           function pointer
-         * @param args          arguments for function
+         * @fn SetKeyPressedCallback
+         * @brief sets function which will be called on key pressed event
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: probably should mansion that function has to take Evnet& as the first parameter.

File: engine/src/core/input/input_manager.h [hunk 11]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
 
         /**
-         * @fn                  SetKeyTypedCallback
-         * @brief               sets function which will be called on key typed event
+         * @fn SetKeyTypedCallback
+         * @brief sets function which will be called on key typed event
          * @param keycode [int] GLFW key id
-         * @param fun           function pointers
-         * @param args          arguments for function
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: same as above

File: engine/src/core/input/input_manager.h [hunk 12]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
         static void SetKeyTypedCallback(F&& fun, Args&& ... args) noexcept;
 
         /**
-         * @fn                  SetKeyReleasedCallback
-         * @brief               sets function which will be called on key released event
-         * @param fun           function pointers
-         * @param args          arguments for function
+         * @fn SetKeyReleasedCallback
+         * @brief sets function which will be called on key released event
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: same as above

File: engine/src/core/input/input_manager.h [hunk 13]
```diff
@@ -56,183 +58,166 @@ namespace Bald::Input {
...
         static void SetKeyReleasedCallback(F&& fun, Args&& ... args) noexcept;
 
         /**
-         * @fn                     SetMouseMovedCallback
-         * @brief                  sets function which will be called on mouse moved event
-         * @param fun              function pointer
-         * @param args             arguments for function
+         * @fn SetMouseMovedCallback
+         * @brief sets function which will be called on mouse moved event
+         * @param [F&&] fun -> function pointer
```
Comment (BIGbadEL) [inline engine/src/core/input/input_manager.h]: same as above

File: engine/src/core/events/mouse_events.h [hunk 14]
```diff
@@ -145,26 +145,36 @@ namespace Bald {
...
+        * @brief Getter which tells you whether or not the key press was a repeated one.
+        * @return [bool] true -> key is in repeated state
+        *                false -> key is in non repeated state
+        */
+
+        [[nodiscard]] bool IsRepeated() const noexcept { return m_IsRepeated; }
 
     private:
         const unsigned m_ButtonCode; /* < We save key code simply as an integer. */
+        const bool m_IsRepeated; /* < Bool which holds repeated state */
```
Comment (BIGbadEL) [inline engine/src/core/events/mouse_events.h:177]: .

File: engine/src/core/events/event_manager.h [hunk 15]
```diff
@@ -45,45 +45,45 @@ namespace Bald {
...
 
         template<class T, class F, class... Args>
         unsigned Subscribe(HandleType type, F&& callback, Args&& ... args);
 
         /**
-        * @fn                    Unsubscribe
-        * @brief                 Templated function which allows you to unsubscribe from any kind of engine's event
-        * @param [unsigned]      id -> Handler id, meaning id of a function you want to get rid of
+        * @fn Unsubscribe
+        * @brief Templated function which allows you to unsubscribe from any kind of engine's event
```
Comment (BIGbadEL) [inline engine/src/core/events/event_manager.h:60]: something about what is T param

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

**Review by BIGbadEL — APPROVED** [review]

**Pull request merged**